kube-runtime/src/watcher.rs

```diff
@@ -102,10 +112,16 @@ impl<K> Event<K> {
     #[must_use]
     pub fn modify(mut self, mut f: impl FnMut(&mut K)) -> Self {
         match &mut self {
-            Event::Applied(obj) | Event::Deleted(obj) => (f)(obj),
+            Event::Applied(obj) | Event::Deleted(obj) => {
+                if let Some(obj) = Arc::get_mut(obj) {
+                    (f)(obj)
+                }
+            }
```

Comment:
As you've noted, this is likely one of the more awkward problems that falls out of this. If this works like you are saying, i.e.

- `watcherstream.modify(f).reflect(writer)` modifies and works
- `watcherstream.reflect(writer).modify(f)` does not modify because the store has a ref to the arc

then this really highlights the fundamental pros and cons this PR is trying to make; we are fundamentally passing the same object around everywhere, and rely on users to clone if they need to (say via [arc's clone-on-write behaviour on Arc::make_mut](https://doc.rust-lang.org/std/sync/struct.Arc.html#method.make_mut)).

Given that reconcilers are already `Arc`d (and they are the main stream-end consumers) I think this is not too bad on the user side, plus it makes reconcilers and direct watch stream consumers the same, which is nice.

For the particular store into modify case, there are two options as i see:

1. make modification after storing an error via  [`Arc::try_unwrap`](https://doc.rust-lang.org/std/sync/struct.Arc.html#method.try_unwrap) - propagating that error from `modify`
2. use `Arc::make_mut` in modify to allow the separation, while pasing the unmodified object around if chosen, having a copy shunted through the remains of the watcher stream.

I personally lead toward 1 because if people have a fn to modify, they can just apply that at the end of their controller stack in a reconciler helper that ultimately does a short term clone, rather than force the clone early on (when the object might be queued for minutes).

There is also a third option; type enforcement of the order (e.g. have reflector modify the output type in a way that makes it incompatible with modify). Although I am not sure how to do that cleanly.